File: snrg_credit_control/snrg_credit_control/report/employee_salary_management_report/employee_salary_management_report.js
frappe.query_reports["Employee Salary Management Report"] = {
  filters: [
    {
      fieldname: "company",
      label: __("Company"),
      fieldtype: "Link",
      options: "Company",
      default: frappe.defaults.get_user_default("Company"),
      reqd: 1,
    },
    {
      fieldname: "month",
      label: __("Month"),
      fieldtype: "Date",
      default: `${frappe.datetime.get_today().slice(0, 7)}-01`,
      reqd: 1,
    },
    {
      fieldname: "employee",
      label: __("Employee"),
      fieldtype: "Link",
      options: "Employee",
    },
    {
      fieldname: "salary_account",
      label: __("Salary Account"),
      fieldtype: "Link",
      options: "Account",
      get_query() {
        return {
          filters: {
            company: frappe.query_report.get_filter_value("company"),
            is_group: 0,
          },
        };
      },
    },
    {
      fieldname: "basic_report",
      label: __("Basic Report"),
      fieldtype: "Check",
      default: 0,
    },
  ],

  onload(report) {
    report.page.add_action_item(__("Export Ready PDF"), () => {
      if (typeof report.pdf_report !== "function" || !frappe.ui.get_print_settings) {
        frappe.msgprint({
          title: __("PDF Export Unavailable"),
          message: __("Please use Menu > PDF for this report."),
          indicator: "orange",
        });
        return;
      }

      const dialog = frappe.ui.get_print_settings(
        false,
        (print_settings) => {
          print_settings.columns = [];
          print_settings.include_filters = 0;
          report.pdf_report(print_settings);
        },
        report.report_doc && report.report_doc.letter_head,
        report.get_visible_columns ? report.get_visible_columns() : [],
        true
      );

      if (report.add_portrait_warning) {
        report.add_portrait_warning(dialog);
      }
    });
  },

  get_pdf_format(report) {
    const context = getSalaryReportPdfContext(report);

    return `
      <style>
        .salary-certified-report {
          color: #111827;
          font-family: Arial, sans-serif;
          font-size: 10.5px;
          line-height: 1.35;
        }
        .salary-certified-report .report-topline {
          border: 1px solid #111827;
          padding: 6px 10px;
          text-align: center;
          font-size: 10px;
          font-weight: 700;
          letter-spacing: .4px;
          text-transform: uppercase;
        }
        .salary-certified-report .company-block {
          border: 1px solid #111827;
          border-top: 0;
          padding: 14px 18px 12px;
          text-align: center;
        }
        .salary-certified-report .company-name {
          margin: 0 0 5px;
          font-size: 18px;
          font-weight: 700;
          text-transform: uppercase;
        }
        .salary-certified-report .company-address {
          color: #374151;
          font-size: 10.5px;
          white-space: pre-line;
        }
        .salary-certified-report .report-title {
          margin: 14px 0 0;
          font-size: 15px;
          font-weight: 700;
          text-transform: uppercase;
        }
        .salary-certified-report .meta-table,
        .salary-certified-report .salary-table {
          width: 100%;
          border-collapse: collapse;
        }
        .salary-certified-report .meta-table {
          margin: 12px 0 14px;
          border: 1px solid #111827;
        }
        .salary-certified-report .meta-table td {
          border: 1px solid #111827;
          padding: 6px 8px;
          vertical-align: top;
          width: 25%;
        }
        .salary-certified-report .meta-label {
          color: #4b5563;
          display: block;
          font-size: 9px;
          font-weight: 700;
          margin-bottom: 2px;
          text-transform: uppercase;
        }
        .salary-certified-report .meta-value {
          font-size: 11px;
          font-weight: 700;
        }
        .salary-certified-report .salary-table {
          border: 1px solid #111827;
          margin-top: 8px;
        }
        .salary-certified-report .salary-table th,
        .salary-certified-report .salary-table td {
          border: 1px solid #111827;
          padding: 6px 7px;
          vertical-align: top;
        }
        .salary-certified-report .salary-table th {
          background: #f3f4f6;
          color: #111827;
          font-weight: 700;
          text-align: left;
        }
        .salary-certified-report .salary-table .sr-col {
          text-align: center;
          width: 42px;
        }
        .salary-certified-report .salary-table .number-cell {
          text-align: right;
          white-space: nowrap;
        }
        .salary-certified-report .salary-table tfoot td {
          background: #f9fafb;
          font-weight: 700;
        }
        .salary-certified-report .certification {
          border: 1px solid #111827;
          margin-top: 16px;
          padding: 10px 12px;
        }
        .salary-certified-report .certification-title {
          font-weight: 700;
          margin-bottom: 4px;
          text-transform: uppercase;
        }
        .salary-certified-report .signature-row {
          margin-top: 32px;
          width: 100%;
        }
        .salary-certified-report .signature-cell {
          display: inline-block;
          width: 49%;
          vertical-align: bottom;
        }
        .salary-certified-report .signature-line {
          border-top: 1px solid #111827;
          display: inline-block;
          min-width: 190px;
          padding-top: 5px;
        }
      </style>

      <div class="salary-certified-report">
        <div class="report-topline">Certified Salary Statement</div>
        <div class="company-block">
          <h1 class="company-name">${context.company_name}</h1>
          <div class="company-address">${context.company_address}</div>
          <div class="report-title">Employee Salary Management Report</div>
        </div>

        <table class="meta-table">
          <tr>
            <td>
              <span class="meta-label">Salary Month</span>
              <span class="meta-value">${context.month_label}</span>
            </td>
            <td>
              <span class="meta-label">Period</span>
              <span class="meta-value">${context.period_label}</span>
            </td>
            <td>
              <span class="meta-label">Report Type</span>
              <span class="meta-value">${context.report_type}</span>
            </td>
            <td>
              <span class="meta-label">Generated On</span>
              <span class="meta-value">${context.generated_on}</span>
            </td>
          </tr>
        </table>

        <table class="salary-table">
          <thead>
            <tr>
              <th class="sr-col">S. No.</th>
              {% for (let column of columns) { %}
                <th class="{%= column.fieldtype === "Currency" ? "number-cell" : "" %}">
                  {%= frappe.utils.escape_html(__(column.label || "")) %}
                </th>
              {% } %}
            </tr>
          </thead>
          <tbody>
            {% for (let i = 0; i < data.length; i++) { %}
              {% let row = data[i]; %}
              <tr>
                <td class="sr-col">{%= i + 1 %}</td>
                {% for (let column of columns) { %}
                  {% let value = frappe.format(row[column.fieldname], column, { inline: true }, row); %}
                  <td class="{%= column.fieldtype === "Currency" ? "number-cell" : "" %}">
                    {%= value || "" %}
                  </td>
                {% } %}
              </tr>
            {% } %}
          </tbody>
          <tfoot>
            <tr>
              <td colspan="{%= Math.max(columns.length, 1) %}">Total Salary Amount</td>
              <td class="number-cell">{%= format_currency(data.reduce((total, row) => total + flt(row.salary_amount), 0), data[0] && data[0].currency ? data[0].currency : "INR") %}</td>
            </tr>
          </tfoot>
        </table>

        <div class="certification">
          <div class="certification-title">Certification</div>
          <div>
            Certified that the salary details stated above are generated from the company's posted accounting records for the selected salary month.
          </div>
          <div class="signature-row">
            <div class="signature-cell">
              <span class="signature-line">Prepared By</span>
            </div>
            <div class="signature-cell" style="text-align:right;">
              <span class="signature-line">Authorised Signatory</span>
            </div>
          </div>
        </div>
      </div>
    `;
  },

  formatter(value, row, column, data, default_formatter) {
    value = default_formatter(value, row, column, data);

    if (column.fieldname === "salary_amount" && data && Number(data.salary_amount || 0) > 0) {
      return `<span style="font-weight:600;">${value}</span>`;
    }

    if (column.fieldname === "voucher_no" && data && data.voucher_no) {
      return `<a href="/app/journal-entry/${encodeURIComponent(data.voucher_no)}">${value}</a>`;
    }

    return value;
  },
};

function getSalaryReportPdfContext(report) {
  const filterValues = report.get_filter_values ? report.get_filter_values() : {};
  const monthDate = filterValues.month || (report.get_filter_value && report.get_filter_value("month"));
  const monthRange = getSalaryReportMonthRange(monthDate);
  const header = report.$status && report.$status.find(".salary-report-print-header");
  const companyName = header && header.find("h2").text()
    ? header.find("h2").text()
    : filterValues.company || (report.get_filter_value && report.get_filter_value("company")) || "";
  const companyAddress = header && header.find(".company-address").text()
    ? header.find(".company-address").text()
    : "";
  const generatedOn = frappe.datetime && frappe.datetime.now_datetime
    ? frappe.datetime.str_to_user(frappe.datetime.now_datetime())
    : new Date().toLocaleString();

  return {
    company_name: escapeSalaryReportHtml(companyName),
    company_address: escapeSalaryReportHtml(companyAddress),
    month_label: escapeSalaryReportHtml(getSalaryReportMonthLabel(monthDate)),
    period_label: escapeSalaryReportHtml(
      monthRange.from_date && monthRange.to_date
        ? `${frappe.datetime.str_to_user(monthRange.from_date)} to ${frappe.datetime.str_to_user(monthRange.to_date)}`
        : ""
    ),
    report_type: filterValues.basic_report ? "Basic Report" : "Detailed Report",
    generated_on: escapeSalaryReportHtml(generatedOn),
  };
}

function getSalaryReportMonthRange(monthDate) {
  if (!monthDate) {
    return { from_date: "", to_date: "" };
  }

  const parts = monthDate.split("-");
  if (parts.length < 2) {
    return { from_date: monthDate, to_date: monthDate };
  }

  const year = Number(parts[0]);
  const month = Number(parts[1]);
  const lastDay = new Date(year, month, 0).getDate();
  const paddedMonth = String(month).padStart(2, "0");

  return {
    from_date: `${year}-${paddedMonth}-01`,
    to_date: `${year}-${paddedMonth}-${String(lastDay).padStart(2, "0")}`,
  };
}

function getSalaryReportMonthLabel(monthDate) {
  if (!monthDate) {
    return "";
  }

  const parts = monthDate.split("-");
  if (parts.length < 2) {
    return monthDate;
  }

  const date = new Date(Number(parts[0]), Number(parts[1]) - 1, 1);
  return date.toLocaleString(undefined, { month: "long", year: "numeric" });
}

function escapeSalaryReportHtml(value) {
  return frappe.utils.escape_html(String(value || ""));
}

File: snrg_credit_control/snrg_credit_control/report/employee_salary_management_report/employee_salary_management_report.py
from collections import defaultdict
from datetime import datetime

import frappe
from frappe import _
from frappe.utils import cint, flt, formatdate, get_first_day, get_last_day, getdate


def execute(filters=None):
    filters = frappe._dict(filters or {})
    validate_filters(filters)

    from_date, to_date = get_month_range(filters.month)
    data = get_data(filters, from_date, to_date)
    message = get_print_header(filters.company, from_date, to_date)
    report_summary = get_report_summary(data)

    return get_columns(filters), data, message, None, report_summary


def validate_filters(filters):
    if not filters.get("company"):
        frappe.throw(_("Please select a Company."))
    if not filters.get("month"):
        frappe.throw(_("Please select a Month."))


def get_columns(filters):
    if filters.get("basic_report"):
        return [
            {
                "label": _("Employee Name"),
                "fieldname": "employee_name",
                "fieldtype": "Data",
                "width": 280,
            },
            {
                "label": _("Salary Amount"),
                "fieldname": "salary_amount",
                "fieldtype": "Currency",
                "options": "currency",
                "width": 150,
            },
        ]

    return [
        {
            "label": _("Sr No"),
            "fieldname": "sr_no",
            "fieldtype": "Int",
            "width": 60,
        },
        {
            "label": _("Posting Date"),
            "fieldname": "posting_date",
            "fieldtype": "Date",
            "width": 105,
        },
        {
            "label": _("Employee"),
            "fieldname": "employee",
            "fieldtype": "Link",
            "options": "Employee",
            "width": 140,
        },
        {
            "label": _("Employee Name"),
            "fieldname": "employee_name",
            "fieldtype": "Data",
            "width": 220,
        },
        {
            "label": _("Salary Amount"),
            "fieldname": "salary_amount",
            "fieldtype": "Currency",
            "options": "currency",
            "width": 135,
        },
        {
            "label": _("Net Payable"),
            "fieldname": "net_payable",
            "fieldtype": "Currency",
            "options": "currency",
            "width": 125,
        },
        {
            "label": _("Deductions / Adjustments"),
            "fieldname": "deductions",
            "fieldtype": "Currency",
            "options": "currency",
            "width": 155,
        },
        {
            "label": _("Journal Entry"),
            "fieldname": "voucher_no",
            "fieldtype": "Link",
            "options": "Journal Entry",
            "width": 160,
        },
        {
            "label": _("User Remark"),
            "fieldname": "user_remark",
            "fieldtype": "Data",
            "width": 260,
        },
    ]


def get_data(filters, from_date, to_date):
    values = {
        "company": filters.company,
        "from_date": from_date,
        "to_date": to_date,
        "salary_account": filters.get("salary_account"),
        "employee": filters.get("employee"),
    }

    salary_rows = get_salary_rows(values)
    if not salary_rows:
        return []

    voucher_names = [row.voucher_no for row in salary_rows]
    all_employee_rows = get_employee_rows(voucher_names)
    employee_rows = [
        row
        for row in all_employee_rows
        if not filters.get("employee") or row.employee == filters.get("employee")
    ]
    if not employee_rows:
        return []

    salary_by_voucher = {row.voucher_no: row for row in salary_rows}
    employee_count_by_voucher = defaultdict(int)
    for row in all_employee_rows:
        employee_count_by_voucher[row.voucher_no] += 1

    rows = []
    for employee_row in employee_rows:
        salary_row = salary_by_voucher.get(employee_row.voucher_no)
        if not salary_row:
            continue

        salary_amount = flt(salary_row.salary_amount)
        if employee_count_by_voucher[employee_row.voucher_no] > 1:
            salary_amount = flt(employee_row.net_payable)

        rows.append(
            frappe._dict(
                {
                    "posting_date": salary_row.posting_date,
                    "employee": employee_row.employee,
                    "employee_name": employee_row.employee_name,
                    "salary_amount": salary_amount,
                    "net_payable": flt(employee_row.net_payable),
                    "deductions": salary_amount - flt(employee_row.net_payable),
                    "voucher_no": employee_row.voucher_no,
                    "user_remark": salary_row.user_remark,
                    "currency": salary_row.currency,
                }
            )
        )

    rows.sort(key=lambda row: (row.posting_date, row.employee_name or row.employee or "", row.voucher_no))
    for idx, row in enumerate(rows, start=1):
        row.sr_no = idx

    return rows


def get_salary_rows(values):
    salary_account_condition = "jea.account = %(salary_account)s"
    if not values.get("salary_account"):
        salary_account_condition = """
            jea.account IN (
                SELECT name
                FROM `tabAccount`
                WHERE company = %(company)s
                  AND is_group = 0
                  AND root_type = 'Expense'
                  AND account_name LIKE 'Salary%%'
            )
        """

    return frappe.db.sql(
        f"""
        SELECT
            je.name AS voucher_no,
            je.posting_date,
            je.user_remark,
            je.company,
            company.default_currency AS currency,
            SUM(jea.debit_in_account_currency - jea.credit_in_account_currency) AS salary_amount
        FROM `tabJournal Entry` je
        INNER JOIN `tabJournal Entry Account` jea ON jea.parent = je.name
        INNER JOIN `tabCompany` company ON company.name = je.company
        WHERE je.docstatus = 1
          AND je.company = %(company)s
          AND je.posting_date BETWEEN %(from_date)s AND %(to_date)s
          AND {salary_account_condition}
        GROUP BY je.name, je.posting_date, je.user_remark, je.company, company.default_currency
        HAVING salary_amount > 0
        ORDER BY je.posting_date ASC, je.name ASC
        """,
        values,
        as_dict=True,
    )


def get_employee_rows(voucher_names):
    if not voucher_names:
        return []

    values = {}
    placeholders = []
    for idx, voucher_name in enumerate(voucher_names):
        key = f"voucher_{idx}"
        placeholders.append(f"%({key})s")
        values[key] = voucher_name

    return frappe.db.sql(
        f"""
        SELECT
            jea.parent AS voucher_no,
            jea.party AS employee,
            COALESCE(emp.employee_name, jea.party) AS employee_name,
            SUM(jea.credit_in_account_currency - jea.debit_in_account_currency) AS net_payable
        FROM `tabJournal Entry Account` jea
        LEFT JOIN `tabEmployee` emp ON emp.name = jea.party
        WHERE jea.parent IN ({", ".join(placeholders)})
          AND jea.party_type = 'Employee'
          AND IFNULL(jea.party, '') != ''
        GROUP BY jea.parent, jea.party, emp.employee_name
        HAVING net_payable > 0
        ORDER BY employee_name ASC, employee ASC
        """,
        values,
        as_dict=True,
    )


def get_month_range(month_value):
    value = str(month_value)
    if len(value) == 7:
        parsed = datetime.strptime(value, "%Y-%m").date()
    else:
        parsed = getdate(value)

    return get_first_day(parsed), get_last_day(parsed)


def get_print_header(company, from_date, to_date):
    company_doc = frappe.get_cached_doc("Company", company)
    address = get_company_address(company)
    period = _("{0} to {1}").format(
        formatdate(from_date, "dd-MM-yyyy"),
        formatdate(to_date, "dd-MM-yyyy"),
    )

    address_html = f"<div>{frappe.utils.escape_html(address)}</div>" if address else ""
    return f"""
        <style>
            .salary-report-print-header {{
                margin: 0 0 12px;
                padding: 10px 0 12px;
                border-bottom: 1px solid #d9d9d9;
                text-align: center;
            }}
            .salary-report-print-header h2 {{
                margin: 0 0 4px;
                font-size: 18px;
                font-weight: 700;
            }}
            .salary-report-print-header .company-address {{
                color: #555;
                font-size: 11px;
                line-height: 1.4;
                white-space: pre-line;
            }}
            .salary-report-print-header .period {{
                margin-top: 8px;
                font-size: 13px;
                font-weight: 600;
            }}
            @media print {{
                .salary-report-print-header {{
                    margin-top: -8px;
                }}
            }}
        </style>
        <div class="salary-report-print-header">
            <h2>{frappe.utils.escape_html(company_doc.company_name or company)}</h2>
            <div class="company-address">{address_html}</div>
            <div class="period">{_("Employee Salary Management Report")} - {period}</div>
        </div>
    """


def get_company_address(company):
    address = frappe.db.sql(
        """
        SELECT
            address.address_line1,
            address.address_line2,
            address.city,
            address.state,
            address.pincode,
            address.country
        FROM `tabAddress` address
        INNER JOIN `tabDynamic Link` link
            ON link.parent = address.name
            AND link.parenttype = 'Address'
        WHERE link.link_doctype = 'Company'
          AND link.link_name = %s
          AND IFNULL(address.disabled, 0) = 0
        ORDER BY
            IFNULL(address.is_primary_address, 0) DESC,
            IFNULL(address.is_your_company_address, 0) DESC,
            address.creation ASC
        LIMIT 1
        """,
        (company,),
        as_dict=True,
    )
    if not address:
        return ""

    row = address[0]
    parts = [
        row.address_line1,
        row.address_line2,
        ", ".join(filter(None, [row.city, row.state, row.pincode])),
        row.country,
    ]
    return "\n".join(part for part in parts if part)


def get_report_summary(data):
    total_salary = sum(flt(row.salary_amount) for row in data)
    total_net_payable = sum(flt(row.net_payable) for row in data)
    total_deductions = sum(flt(row.deductions) for row in data)

    return [
        {
            "value": total_salary,
            "label": _("Total Salary"),
            "datatype": "Currency",
            "indicator": "Blue",
        },
        {
            "value": total_net_payable,
            "label": _("Net Payable"),
            "datatype": "Currency",
            "indicator": "Green",
        },
        {
            "value": total_deductions,
            "label": _("Deductions"),
            "datatype": "Currency",
            "indicator": "Orange" if cint(total_deductions) else "Grey",
        },
    ]
